cdc/generation.cc

```diff
@@ -119,176 +123,213 @@ const std::vector<token_range_description>& topology_description::entries() cons
+    /*
+     * Whether a vnode is a wrap around which means it contains both 2^63 - 1
+     * and -2^63.
+     */
+    bool is_wrap_around() const { return start >= end; }
+    /*
+     * Finds the first token in this vnode which is a wrap around.
+     *
+     * Such vnode has to be considered as two ranges.
+     * Range (|start|, 2^63-1] first and only when no token for shard
+     * |shard_idx| is found there, a second range of [-2^63, |end|] is searched.
+     */
+    dht::token find_first_token_for_shard_in_wrap_around(
+            const dht::sharder& sharder, size_t shard_idx) const;
+    /*
+     * Finds the first token in this vnode which is a wrap around.
+     */
+    dht::token find_first_token_for_shard_in_non_wrap_around(
+            const dht::sharder& sharder, size_t shard_idx) const;
+};
+
+dht::token vnode::find_first_token_for_shard(
+        const dht::sharder& sharder, size_t shard_idx) const {
+    return is_wrap_around()
+        ? find_first_token_for_shard_in_wrap_around(sharder, shard_idx)
+        : find_first_token_for_shard_in_non_wrap_around(sharder, shard_idx);
+}
 
-    std::vector<token_range_description> entries(tokens.size());
-    int spots_to_fill = 0;
+static
+dht::token next_token_no_wrap_around(dht::token t) {
+    return dht::token::from_int64(dht::token::to_int64(t) + 1);
+}
 
-    for (size_t i = 0; i < tokens.size(); ++i) {
-        auto& entry = entries[i];
-        entry.token_range_end = tokens[i];
+static
+dht::token smallest_token() {
+    return dht::token::from_int64(std::numeric_limits<int64_t>::min());
+}
 
-        if (bootstrap_tokens.count(entry.token_range_end) > 0) {
-            entry.streams.resize(smp::count);
-            entry.sharding_ignore_msb = cfg.murmur3_partitioner_ignore_msb_bits();
-        } else {
-            auto endpoint = token_metadata.get_endpoint(entry.token_range_end);
-            if (!endpoint) {
-                throw std::runtime_error(format("Can't find endpoint for token {}", entry.token_range_end));
-            }
-            auto sc = get_shard_count(*endpoint, gossiper);
-            entry.streams.resize(sc > 0 ? sc : 1);
-            entry.sharding_ignore_msb = get_sharding_ignore_msb(*endpoint, gossiper);
-        }
+static
+dht::token next_token_wrap_around(dht::token t) {
+    return dht::token::to_int64(t) == std::numeric_limits<int64_t>::max()
+            ? smallest_token() : next_token_no_wrap_around(t);
+}
 
-        spots_to_fill += entry.streams.size();
+dht::token vnode::find_first_token_for_shard_in_wrap_around(
+        const dht::sharder& sharder, size_t shard_idx) const {
+    auto t = next_token_wrap_around(start);
+    if (sharder.shard_of(t) != shard_idx) {
+        t = sharder.token_for_next_shard(start, shard_idx);
     }
+    if (!t.is_maximum()) {
+        // This means we have found a token for shard |shard_idx| before 2^63-1
+        return t;
+    }
+    // No token owned by shard |shard_idx| was found in (|start|, 2^63 - 1]
```

Comment:
what if `t == 2^63 - 1` and `t` belongs to `shard_idx`?
Then `t.is_maximum()` is true, and we can return `t`.

Reply:
no - is_maximum return true only for the special kind of token that's after_all_tokens.

Reply:
Would it be possible to completely get rid of those other token kinds that are not represented by integers? What do we need them for?

Reply:
For ranges. I will look at removing them but not in this PR.